### ISSUE TITLE
HashMap does not remove a key

### ISSUE THREAD

**Issue by iamim** [louthy/language-ext#624]
It seems like the problems with `TrieMap` implementation persist. I just observed a case where HashMap would not remove a key with both `Remove`. I temporarily fixe dit with `toHashMap(hashMap.AsEnumerable())`.
I'll try to replicate it and update the issue.

**Comment by louthy** [comment]
I've created a repro, working on it now.

**Comment by louthy** [comment]
`v3.3.14` fixes the issue.

I have been running this test for the past hour:
```c#
    static void HashMapRemoveTest()
    {
        var cnt = 10000;
        var rnd = new Random();
        var tries = 1;
        while(true)
        {
            Console.WriteLine("Try: " + tries);

            Seq<int> xs = default;
            HashMap<int, int> hm = default;

            for(var i = 0; i < cnt; i++)
            {
                var n = rnd.Next();
                if(!hm.ContainsKey(n))
                {
                    hm = hm.Add(n, n);
                    xs = xs.Add(n);
                }
            }

            foreach(var x in xs)
            {
                hm = hm.Remove(x);
                if(hm.ContainsKey(x))
                {
                    throw new Exception();
                }
            }
            tries++;
        }
    }
```
Also with varying levels of `cnt`.  I've run many 100,000s of tests with random data-sets and everything passes.  So I'm confident in this fix.

**Comment by louthy** [comment]
Running this test now and all is good.  It randomises the count as well as does random adds after removing to replicate real world usage:
```c#
    static void HashMapRemoveTest()
    {
        var rnd = new Random();
        var tries = 1;
        while(true)
        {
            var cnt = 1 + Math.Abs(rnd.Next() % 100000);

            Console.WriteLine($"Try: {tries} - {cnt} items");

            Seq<int> xs = default;
            HashMap<int, int> hm = default;

            for(var i = 0; i < cnt; i++)
            {
                var n = rnd.Next();
                if(!hm.ContainsKey(n))
                {
                    hm = hm.Add(n, n);
                    xs = xs.Add(n);
                }
            }

            foreach(var x in xs)
            {
                hm = hm.Remove(x);
                if(hm.ContainsKey(x))
                {
                    throw new Exception();
                }

                // Add something else to make sure add works after remove
                if (Math.Abs(rnd.Next() % 1000) < 100)
                {
                    var n = rnd.Next();
                    if (!hm.ContainsKey(n))
                    {
                        hm = hm.Add(n, n);
                    }
                }
            }
            tries++;
        }
    }
```

**Comment by iamim** [comment]
Awesome, thank you!
I'll run some property tests for my solution and close the issue if it's ok.

**Comment by iamim** [comment]
Unfortunately, I found a case when one map contained the same key twice. It's rather hard to replicate because I'm running a bunch of randomized tests as an end-to-end test for the pure part of the program.
I use many functions from the `HashMap` interface and I suspect that `TrieMap` state is corrupted by one of those.

**Comment by louthy** [comment]
Yep, I'm sure there's a deeper issue with the bit-mask, but hoped this more brute force fix would be a good temporary fix (I'm going to be a away for a few weeks soon, so wanted to get this out). 

The test I built above was to allow me to get at the data causing the problem (breakpoint on the `throw`, so that'd be the best way to get a repro step here).

I'll try and do some more digging tomorrow, but might run out of time.

**Comment by iamim** [comment]
Alright, thanks!
I'm trying to introduce functional programming in our backend and this issue has worried me a lot. I somehow forgot about the possibility of downgrading, hah. 3.1.24 seems to works just fine.
I had some ideas on how to generate tests for an implementation and prove that's identical to an existing implementation, e.g. `HashMap` vs .NET `Dictionary`. Will try my best to help if you are not able to find the bug before leaving.

**Comment by louthy** [comment]
I've got a test now that proves the `Remove` is causing a problem, so I should be able to narrow it down pretty quickly. 

This throws as expected:
```c#
        HashMap<int, int> items = default;

        items = items.Add(2085595311, 2085595311);
        items = items.Add(2085595311, 2085595311);
```
This doesn't:
```c#
        HashMap<int, int> items = default;

        items = items.Add(2085595311, 2085595311);
        items = items.Remove(2085595311);
        items = items.Add(2085595311, 2085595311);
        items = items.Add(2085595311, 2085595311);
```

**Comment by louthy** [comment]
Ok, I found the core issue.  It's fixed in `3.3.16` if you want to run it through your tests.

Essentially as the trie grows the mask for each entry has a different value; `2085595311` has a mask of `32768` at level 1, but when there's multiple entries on level one with the same mask then a new level is created and the two items move to that level.  `2085595311` then has a mask of `32`.  

When removing one of the items so there's only one left then there was a check for this so that level 2 gets removed and the one item that's left moves back to level 1.  What wasn't happening was a recalculation of the mask from `32` back to `32768`.  Which meant subsequent checks to see if the item exists would fail and so `2085595311` could be added again.

Because of this I've been able to remove my other temporary fixes, because I suspect this is the only issue.   

The joys of highly optimised imperative programming _/sarcasm_
